File: Java/O(N)/ON_132.java
// Source: https://www.geeksforgeeks.org/maximum-equilibrium-sum-in-an-array/

// Java program to find maximum equilibrium sum.
import java.io.*;

public class ON_132 {
	

	// Function to find maximum
	// equilibrium sum.
	static int findMaxSum(int []arr, int n)
	{
		
		// Array to store prefix sum.
		int []preSum = new int[n];
	
		// Array to store suffix sum.
		int []suffSum = new int[n];
	
		// Variable to store maximum sum.
		int ans = Integer.MIN_VALUE;
	
		// Calculate prefix sum.
		preSum[0] = arr[0];
		for (int i = 1; i < n; i++)
			preSum[i] = preSum[i - 1] + arr[i];
	
		// Calculate suffix sum and compare
		// it with prefix sum. Update ans
		// accordingly.
		suffSum[n - 1] = arr[n - 1];
		
		if (preSum[n - 1] == suffSum[n - 1])
			ans = Math.max(ans, preSum[n - 1]);
			
		for (int i = n - 2; i >= 0; i--)
		{
			suffSum[i] = suffSum[i + 1] + arr[i];
			
			if (suffSum[i] == preSum[i])
				ans = Math.max(ans, preSum[i]);
		}
	
		return ans;
	}
	
	// Driver Code
	static public void main (String[] args)
	{
		int []arr = { -2, 5, 3, 1, 2, 6, -4, 2 };
		int n = arr.length;
		
		System.out.println( findMaxSum(arr, n));
	}
}

// This code is contributed by anuj_67

File: Python/O(N)/on_132.py
# Source: https://www.geeksforgeeks.org/equilibrium-index-of-an-array/

# Python program for the above approach
def find(arr, n):
	mid = n // 2;
	leftSum = 0;
	rightSum = 0;

	# calculation sum to left of mid
	for i in range(mid):
		leftSum += arr[i];
	
	# calculating sum to right of mid
	for i in range(n - 1, mid, -1):
		rightSum += arr[i];
	

	# if rightsum > leftsum
	if (rightSum > leftSum):
	
		# we keep moving right until rightSum become equal or less than leftSum
		while (rightSum > leftSum and mid < n - 1):
			rightSum -= arr[mid + 1];
			leftSum += arr[mid];
			mid += 1;
		
	else:
		# we keep moving right until leftSum become equal or less than RightSum
		while (leftSum > rightSum and mid > 0):
			rightSum += arr[mid];
			leftSum -= arr[mid - 1];
			mid -= 1;
		
	# check if both sum become equal
	if (rightSum == leftSum):
		print("First Point of equilibrium is at index =" , mid);
		return;
	print("First Point of equilibrium is at index =" , -1);

# Driver code
if __name__ == '__main__':
	arr = [ 1, 1, 1, -1, 1, 1, 1 ];
	n = len(arr);
	find(arr, n);

# This code is contributed by gauravrajput1
